src/osp/Universe.h

```diff
@@ -122,27 +126,31 @@ class Universe
      * @tparam ARGS_T Arguments to forward to TYPESAT_T's constructor
      * @return reference to new TYPESAT_T just created
      */
-    template<typename TYPESAT_T, typename ... ARGS_T>
-    TYPESAT_T& type_register(ARGS_T&& ... args);
+    template<typename TYPESAT_T>
+    TypeSatIndex sat_type_register();
 
     /**
-     * Tries to locate an element in the map of registered Satellites by name.
-     * @param name [in] Name of ITypeSatellite
-     * @return Map iterator directly from std::map::find()
+     * @return Names of registered satellites. Access using a TypeSatIndex
      */
-    MapSatType::iterator sat_type_find(const std::string &name)
-    {
-        return m_satTypes.find(name);
-    }
+    constexpr std::vector<std::string_view> const& sat_type_get_names()
+    { return m_typeSatNames; }
 
     /**
-     * Tries to locate an element in the map of registered Satellite by type,
-     * and casts it to SATTYPE_T
-     * @tparam SATTYPE_T Type of ITypeSatellite to find
-     * @return Reference to satellite type found
+     * Find index of a registered satellite by name
+     * @param name [in] Name of satellite to find
+     * @return Index of type, TypeSatIndex::Invalid if not found.
+     */
+    TypeSatIndex sat_type_find_index(std::string_view name);
+
+    /**
+     * Find index of a registered satellite by type. The type should contain a
+     * static member string named smc_name
+     * @tparam SATTYPE_T Satellite type containing smc_name
+     * @return Index of type, TypeSatIndex::Invalid if not found.
      */
     template<typename SATTYPE_T>
-    SATTYPE_T& sat_type_find();
+    TypeSatIndex sat_type_find_index()
+    { return sat_type_find_index(SATTYPE_T::smc_name); }
```

Comment:
I'm wondering if it doesn't make sense to use the EnTT type_id system (Or at least the version of it that I'm reading about in 3.6) to do the name lookup.

It has a name() member function.

Reply:
Note; We can manually specialize the type_id for known types, and I believe we can make the ids use the scheme that EnTT defaults to even while specializing it.

That consolidates the type names to a single place, right?

Reply:
If I remember correctly, it uses a few non-standard compiler features, and can return different values. This is something that should be consistent across all clients in a multiplayer scenario.

Reply:
That's why I'm saying to specialize the type_id for each struct / class that we're currently providing an smc_name.

By doing the specialization, we eliminate compiler compatibility problems, and we also concentrate the strings to a single place without any loss of functionality, *and* we make the places where the strings are used a bit sexier.